### ISSUE TITLE
SSL Error on S3 connection

### ISSUE THREAD

**Issue by iaingray** [fog/fog#3075]
I'm getting an error with S3 at the moment. If I set up a connection (using my key + secret, eu-west-1) such that `connection` is an instance of Fog::Storage::AWS::Real, then do 

`puts connection.directories.to_a`, I get an array of directory objects, as expected.

However, if I try:

`p connection.directories.to_a` or `puts connection.directories`

I get the error details in the below trace, after a long delay.

Any idea why this is happening?
```
/home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/ssl_socket.rb:105:in `connect_nonblock': SSL_connect SYSCALL returned=5 errno=0 state=SSLv2/v3 read server hello A (OpenSSL::SSL::SSLError) (Excon::Errors::SocketError)
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/ssl_socket.rb:105:in `block in initialize'
	from /usr/lib/ruby/2.1.0/timeout.rb:91:in `block in timeout'
	from /usr/lib/ruby/2.1.0/timeout.rb:35:in `block in catch'
	from /usr/lib/ruby/2.1.0/timeout.rb:35:in `catch'
	from /usr/lib/ruby/2.1.0/timeout.rb:35:in `catch'
	from /usr/lib/ruby/2.1.0/timeout.rb:106:in `timeout'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/ssl_socket.rb:101:in `initialize'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:414:in `new'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:414:in `socket'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:126:in `request_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/mock.rb:44:in `request_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/instrumentor.rb:22:in `request_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/base.rb:15:in `request_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/base.rb:15:in `request_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/base.rb:15:in `request_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:269:in `request'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/idempotent.rb:12:in `error_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/base.rb:10:in `error_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/base.rb:10:in `error_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:292:in `rescue in request'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:229:in `request'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/idempotent.rb:12:in `error_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/base.rb:10:in `error_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/base.rb:10:in `error_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:292:in `rescue in request'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:229:in `request'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/idempotent.rb:12:in `error_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/base.rb:10:in `error_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/middlewares/base.rb:10:in `error_call'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:292:in `rescue in request'
	from /home/iain/.gem/ruby/2.1.0/gems/excon-0.38.0/lib/excon/connection.rb:229:in `request'
	from /home/iain/.gem/ruby/2.1.0/gems/fog-1.23.0/lib/fog/xml/sax_parser_connection.rb:35:in `request'
	from /home/iain/.gem/ruby/2.1.0/gems/fog-1.23.0/lib/fog/xml/connection.rb:17:in `request'
	from /home/iain/.gem/ruby/2.1.0/gems/fog-1.23.0/lib/fog/aws/storage.rb:535:in `request'
	from /home/iain/.gem/ruby/2.1.0/gems/fog-1.23.0/lib/fog/aws/requests/storage/get_service.rb:21:in `get_service'
	from /home/iain/.gem/ruby/2.1.0/gems/fog-1.23.0/lib/fog/aws/models/storage/directories.rb:11:in `all'
	from /home/iain/.gem/ruby/2.1.0/gems/fog-core-1.23.0/lib/fog/core/collection.rb:139:in `lazy_load'
	from /home/iain/.gem/ruby/2.1.0/gems/fog-core-1.23.0/lib/fog/core/collection.rb:15:in `empty?'
	from /home/iain/.gem/ruby/2.1.0/gems/fog-core-1.23.0/lib/fog/core/collection.rb:84:in `block in inspect'
	from /home/iain/.gem/ruby/2.1.0/gems/formatador-0.2.5/lib/formatador.rb:92:in `indent'
	from /home/iain/.gem/ruby/2.1.0/gems/fog-core-1.23.0/lib/fog/core/collection.rb:77:in `inspect'
```

**Comment by starbelly** [comment]
I too have this issue, similar anyway. I can upload images, create directories, etc. However, if I try to list directories, it hangs ( way past the read/write_timeout(s) ). I tried to replicate this outside of fog with the aws-s3 gem, but could not. I've also tried manually listing buckets with excon outside of fog ( using the headers that fog generates ) and no problems there either. I did find if you comment out the Timeout.timeout calls in lib/excon/socket.rb and lib/excon/ssl_socket.rb, all works, so it _seems_ like a threading issue and not an ssl issue.

**Comment by geemus** [comment]
@starbelly have you tried adjusting the timeout lengths? Do you have a lot of directories? Are the directories you have mostly in one region or pretty spread out? ie I don't know exactly whats up, so whatever further info you can provide may help narrow it down. Thanks!

**Comment by starbelly** [comment]
@geemus I tried changing timeout lengths, but doesn't have an effect. If I change the timeout lengths to 10 seconds it takes around 7 to 8 minutes for the script to throw an error about SSL ( SSL_connect SYSCALL returned=5 errno=0 state=SSLv2/v3 ). If I change it to 540 seconds, same thing, takes about 6 minutes to puke, but instead throws a connection reset by peer error. In regards to what is in S3, it is one empty directory in a single region. 
 I have a copy of my byebug session if you'd like me to email it to you ( too long to paste here ), though I don't know how useful that would be. It allowed me to see that it hangs in timeout.rb when the thread is joined ( around line 93 in timeout.rb ). Please let me know what else I can do to help solve this problem. In trying to find a solution, I found that most people seem to be saying to stay away from timeout.rb and threads and to use IO.select when you can or something like system-timer ( http://ph7spot.com/musings/system-timer ). I also find this post quite interesting : http://redgetan.cc/understanding-timeouts-in-cruby/ and this http://headius.blogspot.com/2008/02/rubys-threadraise-threadkill-timeoutrb.html . Those posts are old, but I believe still relevant.

Here is a successful request, outside of fog using excon directly: 

Code : 

```
require 'excon'
headers = {
    "Authorization"        => "REDACTED",
    "Host"                 => "s3.amazonaws.com",
    "User-Agent"           => "fog/1.26.0 fog-core/1.27.2",
    "x-amz-content-sha256" => "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "x-amz-date"           => "20150105T220801Z"
  }
conn = Excon.new('https://s3.amazonaws.com:443', :debug => true, :persistent => true)
conn.get(:headers => headers)
```
 Response : 
`
 :body          => "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<ListAllMyBucketsResult xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\"><Owner><ID>de24e59a7fe02cb5a603d0a1cd9a2e64541b8ff5bb553d3ed9417f51b2fa7626</ID><DisplayName>REDACTED</DisplayName></Owner><Buckets><Bucket><Name>skadoosh</Name><CreationDate>2014-12-24T07:49:37.000Z</CreationDate></Bucket></Buckets></ListAllMyBucketsResult>"
`

**Comment by geemus** [comment]
Weird. If you run with `EXCON_DEBUG=true` in fog and directly in excon, any noticeable differences in the format or contents of the request? I'm wondering where the discrepancy is... (feel free to post both and I can try to look if that is easier). Thanks!

**Comment by starbelly** [comment]
Diff of the two requests : 

```
--- fog_out	2015-01-06 15:44:27.000000000 -0600
+++ excon_out	2015-01-06 15:42:51.000000000 -0600
@@ -1,22 +1,22 @@
-t
+excon.request
   :chunk_size          => 1048576
   :ciphers             => "HIGH:!SSLv2:!aNULL:!eNULL:!3DES"
   :connect_timeout     => 60
+  :debug               => true
   :debug_request       => true
   :debug_response      => true
-  :expects             => 200
   :headers             => {
     "Authorization"        => "REDACTED"
     "Host"                 => "s3.amazonaws.com"
     "User-Agent"           => "fog/1.26.0 fog-core/1.27.2"
     "x-amz-content-sha256" => "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
-    "x-amz-date"           => "20150106T214342Z"
+    "x-amz-date"           => "20150105T224847Z"
   }
   :host                => "s3.amazonaws.com"
-  :idempotent          => true
+  :idempotent          => false
   :instrumentor        => Excon::StandardInstrumentor
   :instrumentor_name   => "excon"
-  :method              => "GET"
+  :method              => :get
   :middlewares         => [
     Excon::Middleware::ResponseParser
     Excon::Middleware::Expects
@@ -28,11 +28,10 @@
   :nonblock            => true
   :omit_default_port   => false
   :path                => "/"
-  :persistent          => false
+  :persistent          => true
   :port                => 443
   :query               => nil
   :read_timeout        => 60
-  :response_block      => #<Proc:0x007fabbf4e28a8@/Users/bpaxton/.rvm/gems/ruby-2.2.0/gems/fog-xml-0.1.1/lib/fog/xml/sax_parser_connection.rb:31 (lambda)>
   :retries_remaining   => 4
   :retry_limit         => 4
   :scheme              => "https"
```

 Note that I tried getting the same behavior by passing in `:idempotent => true` and so forth, but could not replicate.

**Comment by geemus** [comment]
I wouldn't expect those to matter particularly. Are the excon versions the same between the two attempts?

**Comment by starbelly** [comment]
Yes, 0.42.1

**Comment by starbelly** [comment]
I tried using IO.select in place of Timeout.timeout and that works, though most of the unit tests for excon do not pass, but that's probably because I'm doing something thupid. There doesn't seem to be good resolution to this issue, you're damned if you do and damned if you don't. The best solution seems to be to set the timeout on the socket itself  via SO_RCVTIMEO and SO_SNDTIMEO, but doesn't seem to work with later versions of Ruby ( https://stackoverflow.com/questions/9853516/set-socket-timeout-in-ruby-via-so-rcvtimeo-socket-option ).  Damned if you do, damned if you don't. This makes me sad :(

**Comment by geemus** [comment]
@starbelly yeah, I've had a lot of issues around sockets/ssl and varying support across ruby versions. I tried to settle in toward a "usually works in most places" kind of approach, though I know it is less than ideal in an increasing number of edges. I guess we could do stuff to do it depending on version, but that gets to be a maintenance mess pretty quick. I don't know beyond that, any further ideas on how we could improve here?

**Comment by starbelly** [comment]
You are right with your approach. I mean, only two people have even talked about the issue on here ( AFAIK ). The way I see it, this isn't an excon or fog problem, this is a ruby problem, that's what makes me sad. I definitely wouldn't be in favor of adding different ops and conditions depending on the OS, etc. You said it right, that would become a maintenance nightmare.  The only ideas I have are to either use IO.select or dig into the C API to set the timeout on the socket itself. It seems like the latter is the safer solution. I think I'll create two branches in excon, play around,  and we'll see what happens!

**Comment by geemus** [comment]
@starbelly yeah, select is used in non-blocking in some cases, but didn't seem to always do the right thing across versions. So some of the timeouts are an extra layer in cases the selects don't play nice. Also, the selects can be tricky because SLOW responses that trickle in won't timeout even though sometimes that might be the behavior you would want. Anyway, hard problem that I'v struggled to improve the situation around.

**Comment by starbelly** [comment]
Well, with that said and since I've already played around with IO.select a few times and learned about the trade-offs that come with that, I'll jump right into playing with C.

**Comment by geemus** [comment]
oof...

**Comment by plribeiro3000** [comment]
closed in favor of fog/fog-aws#9
